torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -3797,6 +3798,90 @@ def test_ddp_comm_hook_logging(self):
             # type if it didn't exist.
             self.assertEqual(ddp_logging_data.get("comm_hook", ""), "")
 
+
+        @unittest.skipIf(
+            BACKEND != "nccl" and BACKEND != "gloo",
+            "Only Nccl & Gloo backend support DistributedDataParallel",
+        )
+        @skip_if_lt_x_gpu(2)
+        def test_ddp_hook_with_optimizer_parity(self):
+            rank = self.rank
+            torch.cuda.set_device(rank)
+            model = LargeNet()
+            sgd_lr = 1e-2
+            ddp_model_with_optimizer_hook = torch.nn.parallel.DistributedDataParallel(
+                copy.deepcopy(model).cuda(),
+                device_ids=[self.rank],
+            )
+            ddp_model_with_allreduce_hook = torch.nn.parallel.DistributedDataParallel(
+                copy.deepcopy(model).cuda(),
+                device_ids=[self.rank],
+            )
+            sgd_allreduce_hook = torch.optim.SGD(
+                ddp_model_with_allreduce_hook.parameters(),
+                lr=sgd_lr
+             )
+            ddp_model_with_no_hook = torch.nn.parallel.DistributedDataParallel(
+                copy.deepcopy(model).cuda(),
+                device_ids=[self.rank],
+            )
+            sgd_no_hook = torch.optim.SGD(
+                ddp_model_with_no_hook.parameters(),
+                lr=sgd_lr
+            )
+            allreduce_hook = default.allreduce_hook
+            opt_hook_state = default.OptimizerHookState(
+                _FunctionalSGD,
+                sgd_lr,
+            )
+            ddp_model_with_optimizer_hook.register_comm_hook(
+                None,
+                default.hook_then_optimizer(allreduce_hook, opt_hook_state),
+            )
```

Comment:
Could we wrap these codes into a higher level of API for users to use? e.g. ddp_model._use_functional_optimizer( _FunctionalSGD, sgd_lr)

also, how do we prevent users to call optimizer.step() in the loop?

will allreduce_hook increase memory usage and also increase copy btw bucket tensor and tensors owned by gradBucket?

Reply:
@zhaojuanmao Will definitely add better higher level APIs and discuss the design in follow up PRs. Would be good to land this though so we can continue to experiment (running accuracy and QPS flows now) before we support this feature officially.

> also, how do we prevent users to call optimizer.step() in the loop?
There's no way DDP can know about local optimizer AFAIK, so developer would need to know to not init an optimizer and call it in their loop before hand.

> will allreduce_hook increase memory usage and also increase copy btw bucket tensor and tensors owned by gradBucket?

The hook is sort of an interface now and in fact even when hook is not registered, we run a default allreduce hook instead of explicit call to allreduce. I believe @SciPioneer verified comm. hooks don't incur a memory penalty some time ago as well, but I'll run some tests around memory usage here as well.